acquia/acquia_cms
# ACMS-1920: Hide administration menu links that do not have visible child links

**Motivation**
For users with limited permissions (i.e. content editors), many menu items irrelevant to the role still appear in the toolbar. This can be overwhelming so empty top level menu items should have restricted access and not be shown if the user lacks the permission for it.

This is a known problem where an unnecessarily complicated interface needs to be explained/simplified to newer users.

Fixes #[ACMS-1920](https://backlog.acquia.com/browse/ACMS-1920)

**Proposed changes**
The current approach is to add a route parameter to several routes. Then add an access check for routes that have the new parameter. The access check loads the corresponding menu link and its children. If the user does not have access to any of the children, then deny access to the page. If any of the children have the new route parameter, then recursively check them.

**Alternatives considered**
N/A

**Testing steps**

- Login as the site's superuser, and create a new role - name it anything you wish
- On the 'permissions' page for the new role, grant access to 'Use the administration pages and help', 'View the administration theme' and 'Use the toolbar'
- Create a new user for the site, and give this user the newly created role
- log out as the superuser, and login as the new user, with the new role
- Notice that under 'Manage', you still have "Structure" - clicking it will take you to that page that states that you have no administrative items.

**Merge requirements**
- [_Enhancement_] _Major change_, _Minor change_, _Bug_, _Enhancement_, and/or _Chore_ label applied

**Screenshots**
Before:
![Screenshot 2023-09-04 at 13-47-07 test_page Acquia CMS](https://github.com/acquia/acquia_cms/assets/63041696/91a81275-5068-4a4e-a9c6-b89d17aaaadd)
After:
![Screenshot 2023-09-04 at 13-51-14 test_page Acquia CMS](https://github.com/acquia/acquia_cms/assets/63041696/d3926f4d-8fdc-4560-9495-81d9e1fe19b8)
